### PR TITLE
Add type for matching Constraints

### PR DESCRIPTION
**Pull request by willbeason**
Prep work for compiler sharding. This adds a store of the matchers
corresponding to Constraints by-target and by-kind.

We could do constraint matching per-target, but it's better for us to
fail fast rather than partially review objects when we're just going to
return errors. Thus, ConstraintsFor iterates over all targets itself
rather than requiring its caller to.

The triple-nested maps are ugly - I would not need much convincing
to replace them with types.

Signed-off-by: Will Beason <willbeason@google.com>

### CODE REVIEW

**Comment by willbeason** [comment]
My reason for the current implementation as maps-of-maps-of-maps is that this messiness is contained to the package - the type is fully contained in the `client` package and it is unexported. If we were to move this functionality to its own package, or to need this behavior to be extensible somehow, we'd want to use special types for the per-target and per-kind layers.

**Comment by willbeason** [comment]
Fixes #176

File: constraint/pkg/client/matchers.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	if len(targetMatchers) == 0 {
+		delete(c.matchers, key.target)
+	} else {
+		c.matchers[key.target] = targetMatchers
+	}
+}
+
+// RemoveAll removes all Matchers for Constraints with kind.
+// Returns normally if no entry for the kind existed.
+func (c *constraintMatchers) RemoveAll(kind string) {
```
Comment (maxsmythe) [inline constraint/pkg/client/matchers.go:94]: nit: `RemoveKind`? To disambiguate vs. purging all matchers?

File: constraint/pkg/client/matchers.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+// criteria.
+func (c *constraintMatchers) ConstraintsFor(review interface{}) (map[string]map[string][]string, error) {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	result := make(map[string]map[string][]string)
+
+	errs := errors.ErrorMap{}
+
+	for target, targetMatchers := range c.matchers {
```
Comment (maxsmythe) [inline constraint/pkg/client/matchers.go:130]: What happens if a target is broadcasting that it doesn't deal with a given review via `HandleReview`?

https://github.com/open-policy-agent/frameworks/blob/4ee2501883a584d60b5756dc69888c547e867ba1/constraint/pkg/handler/handler.go#L75-L83

Also, `HandleReview()` should be called to preprocess the review to ensure it's in an appropriate format for `Match` (ex. AugmentedReview in Gatekeeper adding the namespace).

**Review by maxsmythe — COMMENTED** [review]
One nit, one concern around where/when the `HandleReview` hook is called.

File: constraint/pkg/client/matchers.go (hunk 1 above)
Comment (willbeason) [reply constraint/pkg/client/matchers.go:94]: Done.

File: constraint/pkg/client/matchers.go (hunk 2 above)
Comment (willbeason) [reply constraint/pkg/client/matchers.go:130]: Done. Client now calls the matching logic per-target.

File: constraint/pkg/client/matchers.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		kindMatchers = make(map[string]constraints.Matcher)
+	}
+
+	kindMatchers[key.name] = matcher
+	t.matchers[key.kind] = kindMatchers
+}
+
+func (t *targetMatchers) Remove(key matcherKey) {
+	kindMatchers := t.matchers[key.kind]
+	if len(kindMatchers) == 0 {
```
Comment (maxsmythe) [inline constraint/pkg/client/matchers.go:157]: Should we do

```
kindMatchers, ok := t.matchers[key.kind]
```

and use `ok` here to check for existence?

That way we have GC regardless of whether the map happens to be empty?

File: constraint/pkg/client/matchers.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+// Returns normally if no entry for the Constraint existed.
+func (c *constraintMatchers) Remove(key matcherKey) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if len(c.matchers) == 0 {
+		return
+	}
+
+	target := c.matchers[key.target]
```
Comment (maxsmythe) [inline constraint/pkg/client/matchers.go:58]: Should we do

```
targe, ok := c.matchers[key.target]
```

and use `ok` here to check for existence?

That way we have GC regardless of whether the map happens to be empty?

File: constraint/pkg/client/matchers.go (hunk 3 above)
Comment (willbeason) [reply constraint/pkg/client/matchers.go]: Done.

File: constraint/pkg/client/matchers.go (hunk 4 above)
Comment (willbeason) [reply constraint/pkg/client/matchers.go]: Done.

File: constraint/pkg/client/matchers_test.go [hunk 5]
```diff
@@ -0,0 +1,644 @@
...
+			name: "remove from empty target",
+			before: &constraintMatchers{
+				matchers: map[string]targetMatchers{
+					"foo": {},
+				},
+			},
+			key: matcherKey{target: "foo", kind: "bar", name: "qux"},
+			want: &constraintMatchers{
+				matchers: map[string]targetMatchers{
+					"foo": {},
```
Comment (maxsmythe) [inline constraint/pkg/client/matchers_test.go:222]: It looks like GC still isn't taking effect here?

**Review by maxsmythe — APPROVED** [review]
LGTM!

**Review by ritazh — APPROVED** [review]
lgtm

**Pull request merged**